src/Build/Evaluation/ItemSpec.cs

```diff
@@ -91,9 +91,7 @@ public override bool IsMatchOnMetadata(IItem item, IEnumerable<string> metadata)
             {
                 foreach (var referencedItem in ReferencedItems)
                 {
-                    var nonEmptyMetadata = metadata.Where(m => !item.GetMetadataValue(m).Equals(string.Empty) || !referencedItem.Item.GetMetadataValue(m).Equals(string.Empty));
-                    if (nonEmptyMetadata.Count() != 0 && nonEmptyMetadata.All(
-                            m => item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m))))
+                    if (metadata.All(m => item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m)) && !item.GetMetadataValue(m).Equals(string.Empty)))
```

Comment:
Can probably just do `return metadata.All(...` without the if else right?

Reply:
if `metadata.All(...` is false, there might be another referencedItem for which it's true, so that early exit might lead to false results.
This entire function could be turned into:
`return ReferencedItems.Any(referencedItem => metadata.All(...));`
if you're happy with that.